configuration_go/k8sutil/v2/providers.go

```diff
@@ -0,0 +1,430 @@
+
+	ServiceProvider
+	ServiceMonitorProvider
+	PersistentVolumeClaimProvider
+	ManifestsProvider
+}
+
+// Container represents a container in a pod.
+// It implements the ContainerProvider interface.
+type Container struct {
+	Name            string
+	Image           string
+	ImageTag        string
+	ImagePullPolicy corev1.PullPolicy
+	Resources       corev1.ResourceRequirements
+	Env             []corev1.EnvVar
+	LivenessProbe   *corev1.Probe
+	ReadinessProbe  *corev1.Probe
+	Args            []string
+	Ports           []corev1.ContainerPort
+	VolumeMounts    []corev1.VolumeMount
+
+	Volumes      []corev1.Volume
+	VolumeClaims []VolumeClaim
+	ServicePorts []corev1.ServicePort
+	MonitorPorts []monv1.Endpoint
+	Manifests    map[string]ManifestProvider
+}
+
+// GetContainer returns a Kubernetes Container.
+func (c *Container) GetContainer() corev1.Container {
+	return corev1.Container{
+		Name:                     c.Name,
+		Image:                    fmt.Sprintf("%s:%s", c.Image, c.ImageTag),
+		ImagePullPolicy:          c.ImagePullPolicy,
+		Resources:                c.Resources,
+		Env:                      c.Env,
+		LivenessProbe:            c.LivenessProbe,
+		ReadinessProbe:           c.ReadinessProbe,
+		Args:                     c.Args,
+		Ports:                    c.Ports,
+		TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
+		VolumeMounts:             c.VolumeMounts,
+	}
+}
+
+// GetVolumes returns the volumes that the container requires.
+func (c *Container) GetVolumes() []corev1.Volume {
+	return c.Volumes
+}
+
+// GetServicePorts returns the ports that the container exposes.
+func (c *Container) GetServicePorts() []corev1.ServicePort {
+	return c.ServicePorts
+}
+
+// GetServiceMonitorEndpoints returns the endpoints to be monitored by Prometheus.
+func (c *Container) GetServiceMonitorEndpoints() []monv1.Endpoint {
+	return c.MonitorPorts
+}
+
+// GetVolumeClaims returns the volume claims that the container requires.
+func (c *Container) GetVolumeClaims() []VolumeClaim {
+	return c.VolumeClaims
+}
+
+// AddConfigMap adds a config map to the container.
+func (c *Container) AddManifest(filename string, mf ManifestProvider) {
+	if c.Manifests == nil {
+		c.Manifests = map[string]ManifestProvider{}
+	}
+
+	c.Manifests[filename] = mf
+}
+
+func (c *Container) MakeManifests() k8sutil.ObjectMap {
+	ret := k8sutil.ObjectMap{}
+
+	for filename, mf := range c.Manifests {
+		ret[filename] = mf.MakeManifest()
+	}
+
+	return ret
+}
+
+// VolumeClaim represents a volume claim.
+type VolumeClaim struct {
+	Name string
+	Spec corev1.PersistentVolumeClaimSpec
+}
+
+// PersistentVolumeClaimProvider is the interface to be implemented by pods that require volume claims.
+type PersistentVolumeClaimProvider interface {
+	GetVolumeClaims() []VolumeClaim
+}
+
+// PodProvider is the interface to be implemented by pods to be used in a StatefulSet or Deployment.
+type PodProvider interface {
+	MakePodSpec() corev1.PodSpec
```

Comment:
Why is the `PodProvider` the only one with a function like this? I have the impression it should be the `ManifestProvider` interface.

Reply:
The Container also has a `GetContainer` function. It does not implement directly the `ManifestProvider` interface because it is not a final object. It is embedded in a StatefulSet here that needs to know what volume claims are expected for example.